Email Comparison: UI improvements to display introductory offers and discounts more accurately

olaseni opened the pull request:
To support the regional pricing work we are currently wrangling in the backend, we need a few pieces in the frontend to ensure that we display discounts and product offers properly.

We now also have a situation where Google Workspace for instance has a few currencies with diverse billing plans and introductory offers with discounts spanning over different periods. We have to represent this more flexibly in the pricing display.

Sometimes the target product also has a regular sale coupon attached and we have to factor that into the display as well.

#### Proposed Changes

* Extend the introductory product offer display to support non-zero discounts. We used to only display zero pricing, optimized for free trials.
* Extend the offer eligibility for email products to include new domains in the cart. 
* Extend the display of offers to dynamically include the interval unit and count.
* Factor in renewal discounts over periods i.e. `transition_after_renewal_count`
* Rename unit exporting symbol `isDomainEligibleForTitanFreeTrial()` to `isDomainEligibleForTitanIntroductoryOffer()` as it more aptly defines its functionality since we need to make a distinction between a domain that is eligible for an introductory offer, and one that is additionally eligible for a free trial ( i.e. zero cost )

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply D85347-code to the backend. Also apply the test instruction to set up your environment. That patch provides simulated billing plans for certain currencies to make it easier to test various introductory offer and sale coupon scenarios.
* Load the stacked email comparison view in your local dev server i.e. http://calypso.localhost:3000/email/[:site] . You can also use [Calypso Live](https://github.com/Automattic/wp-calypso/pull/66153#issuecomment-1201968909) to navigate that view. You can also reach this view by attempting to purchase a new domain, which leads you to the email upsell. Either way, always ensure to use a domain that has not been used to purchase email before.
* The best way to test the price information unit is to have about four users, with four different currencies: 
    * MXN - free trials on Professional Email, discounted introductory offers on Google Workspace
    * PHP - free trials on Google Workspace, discounted introductory offers on Professional Email
    * TRY - no introductory offer
    * USD - default system billing plans
    * You can also mix in a sale coupon


<img width="800" alt="Screenshot 2022-08-14 at 10 16 41 PM" src="https://user-images.githubusercontent.com/277661/184555281-dab85a3d-4e26-4da5-96cd-1dd71e3673ba.png">

<img width="800" alt="Screenshot 2022-08-14 at 10 16 51 PM" src="https://user-images.githubusercontent.com/277661/184555286-541f7233-8746-4ceb-b657-dd8cb2c37779.png">

<img width="800" alt="Screenshot 2022-08-14 at 10 17 04 PM" src="https://user-images.githubusercontent.com/277661/184555296-76d12ba7-116c-436d-9697-0f81a4d6583a.png">

<img width="800" alt="Screenshot 2022-08-14 at 10 17 12 PM" src="https://user-images.githubusercontent.com/277661/184555304-cdd2caba-039b-45d3-b3bc-df87bcc34b08.png">

<img width="800" alt="Screenshot 2022-08-14 at 10 17 33 PM" src="https://user-images.githubusercontent.com/277661/184555307-e9b1f399-88e0-420c-8add-b0f40424e27f.png">

<img width="800" alt="Screenshot 2022-08-14 at 10 17 42 PM" src="https://user-images.githubusercontent.com/277661/184555309-c458334c-99ff-4261-ac38-ff215eb2513d.png">

<img width="800" alt="Screenshot 2022-08-14 at 10 18 06 PM" src="https://user-images.githubusercontent.com/277661/184555315-df228136-dcaf-4969-848a-a6a6d27bcea0.png">

<img width="800" alt="Screenshot 2022-08-14 at 10 18 18 PM" src="https://user-images.githubusercontent.com/277661/184555319-57b9fd4e-deef-4f17-a13f-c5b21596277d.png">




#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

olaseni committed:
First foray

stephanethomas commented on client/my-sites/email/email-providers-comparison/price/google-workspace.tsx:
```diff
@@ -87,9 +90,14 @@ const GoogleWorkspacePrice = ( {
+							intervalCount: introductoryOffer?.interval_count,
+							intervalUnit: introductoryOffer?.interval_unit,
+						},
+					} ) }
```
Since we now support different intervals, we would need to [take that into account](https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-calypso/README.md#pluralization) when localizing this string.

stephanethomas commented on client/my-sites/email/email-providers-comparison/price/google-workspace.tsx:
```diff
@@ -62,7 +62,8 @@ const GoogleWorkspacePrice = ( {
 	const isDiscounted = hasDiscount( product );
-	const isEligibleForFreeTrial = isDomainEligibleForGoogleWorkspaceFreeTrial( domain );
+	const isEligibleForFreeTrial =
+		isDomainInCart || isDomainEligibleForGoogleWorkspaceFreeTrial( domain );
```
I don't think we can do that, here is what I get when I try to purchase Google Workspace with a new domain when the server patch is not applied (so targeting the old billing plans):

![image](https://user-images.githubusercontent.com/594356/182800210-9b9da6da-f2f0-47c4-8c68-3fd3f4679a50.png)

With Titan, we assumed that we would always offer a free trial, which is not the case with Google Workspace.

olaseni replied:
Good catch!

stephanethomas reviewed (changes requested):
I've found a regression and an issue with localization.

olaseni committed:
Cover cases for where the offer is dynamic

olaseni committed:
Introduce function [isDomainEligibleForTitanIntroductoryOffer] and rectify [isDomainEligibleForTitanFreeTrial]

olaseni committed:
Cleanup shared pricing logic. Restyle discount badge.

- Restyle the discount badge for distinction with more background opacity and solid border.
- Build out the common email provider pricing logic into reusable component

olaseni committed:
Revert changes to critical files

olaseni commented:
> I've found a regression and an issue with localization.

The regression has been addressed.

olaseni committed:
Convert exporting `getEligibleTitanDomain()` from JS to TS. Fix tests broken due to renamed function.

olaseni committed:
Remove premature limitation

stephanethomas commented on client/my-sites/email/email-providers-comparison/price/price-information.tsx:
```diff
@@ -73,63 +61,139 @@ const FreeTrialPriceInformation = ( {
+
+	const introductoryOffer = product.introductory_offer;
+	const shouldProrate = introductoryOffer?.should_prorate_when_offer_ends ?? false;
+	const transitionAfterRenewal = Number( introductoryOffer?.transition_after_renewal_count );
```
I think removing the `_count` suffix makes the name of this variable less explicit. What if we used the same name than in the translation arguments instead?

```suggestion
	const numberOfDiscountedRenewals = Number( introductoryOffer?.transition_after_renewal_count );
```

stephanethomas commented:
Pinging @semihakocer to make sure that we're good with how we display all those promotions.

olaseni committed:
More meaningful variable rename

wongasy commented on client/my-sites/email/email-providers-comparison/price/discount-or-offer-terms.scss:
```diff
@@ -0,0 +1,11 @@
+	background-color: rgba( 184, 230, 191, 0.24 );
+	border: solid 1px var( --studio-green-30 );
+	color: var( --studio-green-50 );
+	margin-left: 5px;
```
@olaseni Instead of applying the margin to all discounted badges, we can perhaps do something like the following, so we apply the margin to any non-first badge? 

```
.discount-or-offer-terms__item:not(:first-child) {
    margin-left: 5px;
}
```

Currently, the margin is rendered even when the discounted badge is rendered as the very first item.
<img width="256" alt="Screen Shot 2022-08-15 at 9 47 33 PM" src="https://user-images.githubusercontent.com/104910361/184781364-e317b4d7-5858-4ae0-83b5-b3094b568a01.png">

olaseni replied:
> Currently, the margin is rendered even when the discounted badge is rendered as the very first item.

Great catch. Fixing ...

wongasy commented on client/my-sites/email/email-providers-comparison/price/discount-or-offer-terms.tsx:
```diff
@@ -0,0 +1,95 @@
+			{ isDiscounted && (
+				<div className="discount-or-offer-terms__discount-badge discount-or-offer-terms__item discount-or-offer-terms__discounted_item badge badge--info-green">
+					{ isEligibleForIntroductoryOffer
+						? translate( '+ Extra %(discount)d%% off', {
```
I wonder if we can simplify this by passing `+ Extra` and `Limited time:` as an `arg`

```
translate( '%(label)s%% %(discount)d%% off', {
	args: {
		label: isEligibleForIntroductoryOffer ? 'Extra' : 'Limited time:',
		discount: product?.sale_coupon?.discount,
	},
	comment: "%(label)s describes the offer (it can be 'Extra' when multiple offers are in place or 'Limited time' for a limited time offer) %(discount)d is a numeric discount percentage (e.g. '40')",
} )
```

Or does this complicate things for translation? 

In terms of the actual label itself, I see the reasoning behind saying `+ Extra`, since we are *adding* an extra 20% off of the introductory offer. However, I feel throwing in the plus operator `+` could create confusing at first glance, since typically discounts operates at `-` (minus an amount). To avoid confusion, perhaps having `Extra` (without the `+`) would do.

olaseni replied:
> Or does this complicate things for translation?

IMO yes, but I can ping @dlind1 for a more informed opinion.

- For one, `placeholder`s should be items that are dynamic, like the cost of a product, or a potential domain name purchase.
- The proposed `placeholder`s themselves are not of the same kind. One is an adjective which describes another `placeholder`, while the other contains a colon.

However, since there are just two possibilities then it's just simpler to have two separate instances.

olaseni replied:
> To avoid confusion, perhaps having Extra (without the +) would do.

OK

dlind1 replied:
In this case `label: isEligibleForIntroductoryOffer ? 'Extra' : 'Limited time:'` would mean that `Extra` and `Limited time` won't be translated. It's best to not concatenate strings or pass translatable strings as arguments, so having two strings would be the best solution, e.g.
`translate( 'Extra %(discount)d%% off')` and `translate( 'Limited time: %(discount)d%% off')`

olaseni committed:
Modify styles so that only non-first discount badges have margins

olaseni committed:
Change translate text: `+ Extra` to `Exra`

wongasy commented on client/my-sites/email/email-providers-comparison/price/price-information.tsx:
```diff
@@ -49,19 +45,11 @@ const DiscountPriceInformation = ( { product }: { product: ProductListItem } ):
 					"%(discountedPrice)s and %(standardPrice)s are formatted prices with the currency (e.g. '$5')",
 			} ) }
 
-			{ isGoogleWorkspace( product ) && (
```
I noticed that we are no longer checking for `isGoogleWorkspace()`, which appears to make sense, since this message can be applicable to both Professional Email and Google. I'm just curious to see why this messaging was restricted to Google at in production?

wongasy commented on client/my-sites/email/email-providers-comparison/price/price-information.tsx:
```diff
@@ -73,63 +61,139 @@ const FreeTrialPriceInformation = ( {
+	return (
+		<div className="price-information__free-trial">
+			{ translate(
+				'Try free today - renews at %(standardPrice)s after trial',
```
The presentation for the scenario of Free trial + Discount seems a bit confusing. (Note that I've updated the test data for the billing plan to apply a `1 month` free trial rather than `1 year`, so it's more realistic, I don't think we'd be offering `1 year` trial 😆)

<img width="1062" alt="Screen Shot 2022-08-16 at 2 30 10 PM" src="https://user-images.githubusercontent.com/104910361/184952686-e3639ec7-1ccd-4502-85b0-0016fe2e33f4.png">

My understanding of offering is:
* User pays 0 in the first month (in trial)
* In the second month, once the trial is over, user pays `P56` for rest of the annual subscription (11 months)
* Next renewal, user pays `P70` (no more discount)

The pricing info (text at the bottom) doesn't factor in the 20% off. Perhaps we can update the info to something like:
`Try free today! Then enjoy first year of the subscription for only P56 (regular price P70).` I'm almost inclined to leave out `(regular price P70)` altogether, since the ~~P70~~ already signifies the regular price. And if we already have the regular price available in **Upgrades > Emails**, then users are well informed (after all, they don't have to worry about the regular price until they renew in the following year).

---

This might have been a bug? I see that `shouldProrate` was set to `true` initially, which should surface the message at L80, but at some point `shouldProrate` became `false`, getting us to this message at L90.

<img width="217" alt="Screen Shot 2022-08-16 at 2 56 06 PM" src="https://user-images.githubusercontent.com/104910361/184958903-cfad9ea4-2ebd-4263-b7b9-1c528687395c.png">

wongasy commented:
Regarding the colour of the discount badge, I know we've had some discussions around it in pbLl1t-1iP-p2#comment-1159, with the concern of "red" signifying "warning". That said, in marketing, "red" can actually be positive:

> The color red stands out on most web pages. It invokes passion, excitement, and urgency. If you want your customers to take urgent action on your product (i.e. purchase it, download an eBook etc.) then Red is the right color.

(https://www.verizon.com/business/small-business-essentials/resources/truth-best-worst-call-action-button-colors-website-212506540/)

> This powerful color is associated with excitement, energy, power, fearlessness, and passion. In sales, call-to-action buttons [use](https://www.verizon.com/business/small-business-essentials/resources/truth-best-worst-call-action-button-colors-website-212506540/) red to empower shoppers to convert because it exudes a sense of urgency.

(https://blog.hubspot.com/the-hustle/psychology-of-color)

There's probably a good reason why there's a site called RedFlagDeals.com for deals and bargains :) It'd be interesting to see how we do with a red badge vs. green badge. Perhaps something worth A/B testing.

That said, I don't want this told hold us up from completing this feature, we can go ahead with the green badge, and create a separate PR to enable an experiment with red badges. Or, if we're bold, we can go for red badges :p

wongasy commented on client/my-sites/email/email-providers-comparison/price/price-information.tsx:
```diff
@@ -73,63 +61,139 @@ const FreeTrialPriceInformation = ( {
+	return (
+		<div className="price-information__free-trial">
+			{ translate(
+				'Discounted today - renews at %(standardPrice)s after trial',
```
The messaging would have to be updated here? Users will be renewing at regular price after the discount period is over and not a trial?

<img width="1067" alt="Screen Shot 2022-08-16 at 3 22 49 PM" src="https://user-images.githubusercontent.com/104910361/184964747-acf8954b-bc64-41ae-9c6c-5a0187f31b1b.png">

I'm also a bit confused about this particular scenario here. I see that there's a 1 year discount (amount unspecified), then on top of it, we're applying an extra 20% off? Is that the idea?

fredrikekelund commented:
> Regarding the colour of the discount badge, I know we've had some discussions around it in pbLl1t-1iP-p2#comment-1159, with the concern of "red" signifying "warning".

For reference, I replied to this here pbLl1t-1iP-p2#comment-1243

olaseni commented:
I can't remember what happened here. but I'm close it as it's so dated.
